### PR TITLE
Add documentation about the different embassy abstraction layers

### PR DESCRIPTION
**Pull request by lulf**
The guide demonstrates the functionality offered by each
layer in Embassy, using code examples.

### CODE REVIEW

File: docs/modules/ROOT/pages/layer_by_layer.adoc [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+It's time to use the Embassy capabilities to its fullest. At the core, Embassy has an async excecutor, or a runtime for async tasks if you will. The executor polls a set of tasks (defined at compile time), and whenever a task `blocks`, the executor will run another task, or put the microcontroller to sleep.
+
+[source,rust]
+----
+include::example$layer-by-layer/blinky-async/src/main.rs[]
+----
+
+The async version looks very similar to the HAL version, apart from a few minor details:
+
+* The main entry point is annotated with a different macro and has an async type signature. This macro creates and starts an Embassy runtime instance and launches the main application task. Using the `Spawner` instance, the application may spawn other tasks.
```
Comment (huntc) [inline docs/modules/ROOT/pages/layer_by_layer.adoc:76]: Do we want to target stable Rust or nightly with this doc? Can’t use the macros with stable, right?

File: docs/modules/ROOT/pages/layer_by_layer.adoc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+* The peripheral initialization is done by the main macro, and is handed to the main task.
+* Before checking the button state, the application is awaiting a transition in the pin state (low -> high or high -> low).
+
+When `button.awit_for_any_edge().await` is called, the executor will pause the main task and put the microcontroller in sleep mode. Internally, the Embassy HAL has configured the interrupt handler for the button (in `ExtiButton`), so that whenever an interrupt is raised, the task awaiting the button will be woken up.
```
Comment (huntc) [inline docs/modules/ROOT/pages/layer_by_layer.adoc:80]: Is “awit” a typo?

File: docs/modules/ROOT/pages/layer_by_layer.adoc [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+The overhead of the executor is minimal, and combined with the enormous simplification of the application, `async` is a great fit for embedded.
```
Comment (huntc) [inline docs/modules/ROOT/pages/layer_by_layer.adoc:82]: Is it also worth mentioning here that async permits concurrency such that other tasks get a look-in and can therefore save energy? I see this as one of the largest benefits.

**Review by huntc — APPROVED** [review]
Love it!

File: docs/modules/ROOT/pages/layer_by_layer.adoc (hunk 1 above)
Comment (lulf) [reply docs/modules/ROOT/pages/layer_by_layer.adoc:76]: Yup, macros require nightly, which I think personally is better because of the reduced boilerplate

**Review by Dirbaio — APPROVED** [review]
Awesome! :D

bors r+

**Pull request merged**